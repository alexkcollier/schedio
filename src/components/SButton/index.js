import SButton from './SButton.vue';

export { SButton };

export default { install: Vue => Vue.component(SButton.name, SButton) };
