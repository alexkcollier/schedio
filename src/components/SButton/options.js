export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
