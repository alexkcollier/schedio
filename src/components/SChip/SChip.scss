.chip {
  @include font-size(6);

  align-items: center;
  background-color: color('ice');
  border-radius: border-radius('small');
  display: inline-flex;
  padding: spacing('half') spacing();
  transition: all easing() duration();
  // ensures SVGs and text align as expected
  vertical-align: top;

  &__content {
    align-items: center;
    display: inline-flex;
  }

  &__close {
    border: none;
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
    box-shadow: none;
    // allows the button to have padding while not increasing the box size
    margin: -1 * spacing('half');
    margin-left: spacing('half');
    margin-right: -1 * spacing();
    padding: spacing('half');

    &:active,
    &:not(:disabled):not(:active):hover {
      box-shadow: none !important;
    }
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      background-color: color($color-name);
      color: choose-contrast-color($color-name);

      .chip__close {
        box-shadow: none;

        &:active,
        &:not(:disabled):not(:active):hover {
          box-shadow: none !important;
        }
      }
    }
  }
}
