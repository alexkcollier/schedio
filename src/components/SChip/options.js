import { props } from '@/assets/styles/tokens/dist/tokens.raw.json'

export const colors = [
  ...Object.values(props).reduce((colorList, { category, name }) => {
    if (category === 'background-color') colorList.add(name.split('-')[0])

    return colorList
  }, new Set()),
].sort()
