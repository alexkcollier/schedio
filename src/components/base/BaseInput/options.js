export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
