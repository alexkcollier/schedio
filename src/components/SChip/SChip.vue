<template>
  <!-- eslint-disable vue/require-component-is -->
  <Component
    :is="tag"
    v-show="isActive"
    class="chip"
    :class="classList"
    :aria-hidden="!isActive"
  >
    <div class="chip__content">
      <slot>Chip content</slot>
    </div>

    <SButton
      v-if="isClosable"
      size="small"
      class="chip__close"
      :color="buttonColor"
      icon-left="x"
      icon-only
      :aria-label="closeAriaLabel"
      @click="() => $emit('close', false)"
    />
  </Component>
</template>

<script>
import { SButton } from '@/components/SButton';
import { colors } from './options';
import defaults from '@/utils/defaults';

export default {
  name: 'SChip',

  components: {
    SButton,
  },

  /**
   * Custom v-model binding to `close` event and `isActive`, rather than `input` and `value`
   */
  model: {
    event: 'close',
    prop: 'isActive',
  },

  props: {
    color: {
      default: () => defaults.chipColor,
      type: String,
      validator: (val) => {
        return (
          !val ||
          colors.includes(val) ||
          console.error(`\`color\` must be one of ${colors.join(', ')}`)
        );
      },
    },

    isClosable: {
      default: false,
      type: Boolean,
    },

    isActive: {
      type: Boolean,
      default: true,
    },

    tag: {
      type: String,
      default: 'div',
    },

    closeAriaLabel: {
      type: String,
      default: 'Close chip',
    },
  },

  computed: {
    classList () {
      return {
        [`chip--color-${this.color}`]: this.color,
      };
    },

    buttonColor () {
      return this.color || 'ice';
    },
  },
};
</script>

<style lang="scss">
.chip {
  @include font-size(6);

  align-items: center;
  background-color: color('ice');
  border-radius: border-radius('small');
  display: inline-flex;
  padding: spacing('half') spacing();
  transition: all easing() duration();
  // ensures SVGs and text align as expected
  vertical-align: top;

  &__content {
    align-items: center;
    display: inline-flex;
  }

  &__close {
    border: none;
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
    box-shadow: none;
    // allows the button to have padding while not increasing the box size
    margin: -1 * spacing('half');
    margin-left: spacing('half');
    margin-right: -1 * spacing();
    padding: spacing('half');

    &:active,
    &:not(:disabled):not(:active):hover {
      box-shadow: none !important;
    }
  }

  @each $color-name in $button-colors {
    &--color-#{$color-name} {
      background-color: color($color-name);
      color: choose-contrast-color($color-name);

      .chip__close {
        box-shadow: none;

        &:active,
        &:not(:disabled):not(:active):hover {
          box-shadow: none !important;
        }
      }
    }
  }
}
</style>
