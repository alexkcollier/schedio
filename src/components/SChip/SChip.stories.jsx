import { SChip } from '@/components/SChip';
import { SChipGroup } from '@/components/SChipGroup';
import { SHeading } from '@/components/SHeading';
import { SCallout } from '@/components/SCallout';
import PropList from '@@/docs/components/PropList';
import { boolean, select, text } from '@storybook/addon-knobs';
import { colors } from './options';
import { SContainerColumn } from '@/components/SContainerColumn';
import { SContainerRow } from '@/components/SContainerRow';
import { generateHeading } from '@/utils/stories/render-functions';

export default {
  title: 'Components/Chips/SChip',

  parameters: {
    jest: 'SChip',
  },
};

export const chip = () => ({
  props: {
    props: {
      default: {
        color: select('color', ['', ...colors], '', 'Optional Props'),
        isClosable: boolean('is-closable', false, 'Optional Props'),
        closeAriaLabel: text('close-aria-label', 'Close toast', 'Optional Props'),
      },
    },
    slots: {
      default: {
        defaultSlot: text('default', 'Some content', '$slots'),
      },
    },
  },

  render (h) {
    const { props, slots } = this.$props;

    return (
      <div>
        <SHeading level="1">Chip</SHeading>
        <p>Chips can be used to convey small bits of information.</p>

        <SHeading level="2">Example</SHeading>
        <SContainerRow>
          <SContainerColumn>
            <SChip {...{ props }}>{slots.defaultSlot}</SChip>
          </SContainerColumn>
        </SContainerRow>

        <SContainerRow>
          <SContainerColumn>
            <SCallout type="warning">
              <code>v-model</code> must be set for a <code>is-closable</code> to work.
            </SCallout>
          </SContainerColumn>
        </SContainerRow>

        {SChip.props && <PropList component={SChip} />}
      </div>
    );
  },
});

chip.story = {
  name: 'Chip',

  parameters: {
    options: {
      showPanel: true,
    },
  },
};

export const closableChip = () => ({
  data () {
    return {
      showComponentCode: false,
      chips: {
        chip1: true,
        chip2: true,
        chip3: true,
        chip4: true,
        chip5: true,
      },
    };
  },

  methods: {
    handleClose (evt, chip) {
      this.chips[chip] = evt;
      console.log(`[SChip]: ${chip} is ${evt} now`);
    },
    resetChips () {
      for (const chip in this.chips) {
        if ({}.hasOwnProperty.call(this.chips, chip)) this.chips[chip] = true;
      }
    },
  },

  render (h) {
    return (
      <div>
        <SHeading level="1">Closable Chip</SHeading>
        <p>
          Chips can be closed with <code>is-closable</code> and <code>v-model</code>. Set{' '}
          <code>is-closable</code> to <code>true</code> and bind <code>v-model</code> to the
            property you want to control the chip's visibility. <code>SChip</code> emits a{' '}
          <code>@close</code> event you can listen to when it is closed.
        </p>

        <SHeading level="2">Example</SHeading>
        <SButton
          icon-left={this.showComponentCode ? 'minus-circle' : 'plus-circle'}
          color="night"
          is-outlined
          size="small"
          style="margin-bottom: 1em;"
          onClick={() => {
            this.showComponentCode = !this.showComponentCode;
          }}
        >
          {this.showComponentCode ? 'Hide' : 'Show'} component code
        </SButton>

        {this.showComponentCode && (
          <pre>
            {`\
import { SChip } from 'schedio-design'

export default {
name: 'SomeComponent',

components: { SChip }

data() {
return {
  chips: {...}
}
},

methods: {
handleClose(evt, chip) {
  // do more...
  console.log(\`[SChip]: \${chip} is \${evt} now\`)
}
},

template: \`
<div>
  <SChip
    v-for="(val, chip) in chips"
    v-model="chip"
    is-closable
    @close="handleClose($event, chip)"
  >
    {{ chip }}
  </SChip>
</div>
\`
}
`}
          </pre>
        )}

        <SHeading level="3">Output</SHeading>

        <SChipGroup>
          {Object.keys(this.chips).map(chip => (
            <SChip
              is-closable
              key={chip}
              onClose={evt => this.handleClose(evt, chip)}
              is-active={this.chips[chip]}
              style="text-transform: capitalize;"
            >
              {chip}
            </SChip>
          ))}
        </SChipGroup>

        <pre>{JSON.stringify(this.chips, null, 2)}</pre>

        <SButton
          color="night"
          size="small"
          is-outlined
          onClick={this.resetChips}
          disabled={Object.values(this.chips).every(v => v)}
        >
          Reset Chips
        </SButton>
      </div>
    );
  },
});

closableChip.story = {
  name: 'Closable Chip',

  parameters: {
    options: {
      showPanel: false,
    },
  },
};

export const chipColors = () => ({
  render (h) {
    return (
      <div>
        <SHeading>Chip Colors</SHeading>

        {['', ...colors].map(color => generateChips(h, color))}
      </div>
    );
  },
});

chipColors.story = {
  name: 'Chip Colors',

  parameters: {
    options: {
      showPanel: false,
    },
  },
};

function generateChips (h, color) {
  const chipName = `${color || ''} chip`;

  return h(SContainerRow, [
    h(SContainerColumn, [
      generateHeading(h, { level: 2, content: color || 'Base' }),
      h(SChipGroup, [
        h(SChip, { props: { color } }, chipName),
        h(SChip, { props: { color, isClosable: true } }, `Closable ${chipName}`),
      ]),
    ]),
  ]);
}
