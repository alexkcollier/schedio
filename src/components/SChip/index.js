import SChip from './SChip.vue';

export { SChip };

export default {
  install: Vue => Vue.component(SChip.name, SChip),
};
