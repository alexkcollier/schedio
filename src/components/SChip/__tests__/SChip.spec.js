import { mount, shallowMount } from '@vue/test-utils'
import SChip from '@/components/SChip/SChip.vue'
import SButton from '@/components/SButton/SButton.vue'
import { colors } from '../options'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SChip.vue', () => {
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  let wrapper

  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('base', () => {
    beforeEach(() => {
      wrapper = shallowMount(SChip, {
        slots: {
          default: 'Slot content',
        },
      })
    })

    it('renders correctly', async () => {
      expect(wrapper.html()).toMatchSnapshot()
      expect(await axe(wrapper.html())).toHaveNoViolations()
    })

    it.each(colors)('can have color %s', async (color) => {
      const colorWrapper = shallowMount(SChip, { propsData: { color } })

      expect(errorSpy).not.toHaveBeenCalled()
      expect(colorWrapper.html()).toMatchSnapshot()
      expect(await axe(colorWrapper.html())).toHaveNoViolations()
    })

    it('validates colors', () => {
      shallowMount(SChip, { propsData: { color: 'Not a color' } })
      expect(errorSpy).toHaveBeenCalledTimes(2)
    })

    it('can have its HTML tag set', () => {
      wrapper.setProps({ tag: 'li' })
      expect(wrapper.contains('li')).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('closable', () => {
    beforeEach(() => {
      wrapper = mount(SChip, {
        slots: {
          default: 'Slot content',
        },
        propsData: {
          isClosable: true,
        },
      })
    })

    it('can be closable', async () => {
      expect(wrapper.contains(SButton)).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
      expect(await axe(wrapper.html())).toHaveNoViolations()
    })

    it('emits a close event when the button is clicked', () => {
      const button = wrapper.find(SButton)
      button.trigger('click')
      expect(wrapper.emitted('close')).toBeTruthy()
    })

    it('works with v-model', () => {
      const chips = {
        chip1: true,
        chip2: true,
        chip3: true,
      }
      const closeWrapper = mount({
        name: 'CloseWrapper',
        components: { SChip },
        data () {
          return {
            chips,
          }
        },
        template: `
          <div>
            <SChip
              v-for="(val, chip) in chips"
              :key="chip"
              is-closable="true"
              v-model="chips[chip]"
            >
              {{ chip }}
            </SChip>
          </div>
        `,
      })

      const chipComponents = closeWrapper.findAll(SChip)
      const buttons = closeWrapper.findAll(SButton)
      const chipNames = Object.keys(chips)
      const totalChips = chipNames.length

      expect(chipComponents.length).toBe(totalChips)
      expect(buttons.length).toBe(totalChips)

      const randIdx = Math.floor(totalChips * Math.random())
      const randKey = chipNames[randIdx]

      buttons.at(randIdx).trigger('click')

      expect(chipComponents.at(randIdx).isVisible()).toBe(false)
      expect(closeWrapper.vm.chips[randKey]).toBe(false)
    })

    it('sets `aria-hidden` on close', async () => {
      wrapper = mount({
        components: { SChip },
        data: () => ({ chip: true }),
        template: '<SChip is-closable="true" v-model="chip">Chip</SChip>',
      })
      wrapper.find(SButton).trigger('click')
      expect(wrapper.find(SChip).attributes('aria-hidden')).toBe('true')
      expect(wrapper.html()).toMatchSnapshot()
      expect(await axe(wrapper.html())).toHaveNoViolations()
    })

    it("can change the button's `aria-label`", async () => {
      wrapper.setProps({ closeAriaLabel: 'Changed' })
      expect(wrapper.find('button').attributes('aria-label')).toBe('Changed')
      expect(wrapper.html()).toMatchSnapshot()
      expect(await axe(wrapper.html())).toHaveNoViolations()
    })
  })
})
