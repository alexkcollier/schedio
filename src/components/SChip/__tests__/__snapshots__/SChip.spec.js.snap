// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SChip.vue base can have color gold 1`] = `
<div class="chip chip--color-gold">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color green 1`] = `
<div class="chip chip--color-green">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color grey 1`] = `
<div class="chip chip--color-grey">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color ice 1`] = `
<div class="chip chip--color-ice">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color magenta 1`] = `
<div class="chip chip--color-magenta">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color night 1`] = `
<div class="chip chip--color-night">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color red 1`] = `
<div class="chip chip--color-red">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color spartan_blue 1`] = `
<div class="chip chip--color-spartan_blue">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color turquoise 1`] = `
<div class="chip chip--color-turquoise">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have color white 1`] = `
<div class="chip chip--color-white">
  <div class="chip__content">Chip content</div>
  <!---->
</div>
`;

exports[`SChip.vue base can have its HTML tag set 1`] = `
<li class="chip">
  <div class="chip__content">Slot content</div>
  <!---->
</li>
`;

exports[`SChip.vue base renders correctly 1`] = `
<div class="chip">
  <div class="chip__content">Slot content</div>
  <!---->
</div>
`;

exports[`SChip.vue closable can be closable 1`] = `
<div class="chip">
  <div class="chip__content">Slot content</div> <button aria-label="Close chip" class="button chip__close button--color-ice button--size-small button--icon-only">
    <!---->
    <sicon icon="x" aria-hidden="true" class="button__icon button__icon--left"></sicon>
    <!---->
    <!---->
  </button>
</div>
`;

exports[`SChip.vue closable can change the button's \`aria-label\` 1`] = `
<div class="chip">
  <div class="chip__content">Slot content</div> <button aria-label="Changed" class="button chip__close button--color-ice button--size-small button--icon-only">
    <!---->
    <sicon icon="x" aria-hidden="true" class="button__icon button__icon--left"></sicon>
    <!---->
    <!---->
  </button>
</div>
`;

exports[`SChip.vue closable sets \`aria-hidden\` on close 1`] = `
<div class="chip" aria-hidden="true" style="display: none;">
  <div class="chip__content">Chip</div> <button aria-label="Close chip" class="button chip__close button--color-ice button--size-small button--icon-only">
    <!---->
    <sicon icon="x" aria-hidden="true" class="button__icon button__icon--left"></sicon>
    <!---->
    <!---->
  </button>
</div>
`;
